### PR TITLE
Follow

### PR DESCRIPTION
**Pull request by kyleasmith1**
Added join/unjoin functionality

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/JoinServlet.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        User user = (User) request.getAttribute(User.USER_ENTITY_NAME);
+        PreparedQuery roomResults = datastore.prepare(new Query(Room.ROOM_ENTITY_NAME));
+        Room room = null;
+
+        for(Entity entity : roomResults.asIterable()) {
+            if (new Room(entity).getRoomKey().getId() == Long.parseLong(request.getParameter(JoinServlet.ROOM_ID_PROPERTY_KEY))) {
+                room = new Room(entity);
+            }
+        }
```
Comment (707090) [inline capstone/src/main/java/com/google/sps/servlets/JoinServlet.java]: This should not query every single room, it should use a query filter

File: capstone/src/main/java/com/google/sps/servlets/JoinServlet.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/join")
+public class JoinServlet extends HttpServlet {
+
+    private static final String JOIN_PROPERTY_KEY = "join";
+    private static final String UNJOIN_PROPERTY_KEY = "unjoin";
+    private static final String ROOM_ID_PROPERTY_KEY = "room_id";
+    private static final String ACTION_PROPERTY_KEY = "action";
```
Comment (707090) [inline capstone/src/main/java/com/google/sps/servlets/JoinServlet.java]: These are not property keys. The last one is a query parameter, the room_id one is a property key, and the first two are just action values. Since the join action can only have one of a set number of possible values, use an enum

File: capstone/src/main/webapp/dashboard-loader.js [hunk 3]
```diff
@@ -28,15 +28,20 @@ function createRoomDivElement(room) {
...
+                    <div class="btn btn-default" id="room-link"></button>
                 </div>
             </div>
             `, "text/html");
+    
+    doc.getElementById("room-title").innerText = room.entity.propertyMap.title;
+    doc.getElementById("room-link").addEventListener("click", function() {
+        window.location.href = "lesson.html?room_id=" + room.entity.key.id;
+    });
+    doc.getElementById("room-link").innerText = "View";
```
Comment (707090) [inline capstone/src/main/webapp/dashboard-loader.js]: This sanitization is not related to following a room and should be part of a separate PR for handling sanitizaiton

File: capstone/src/main/webapp/dashboard-loader.js [hunk 4]
```diff
@@ -28,15 +28,20 @@ function createRoomDivElement(room) {
...
                 <div class="card-body text-center">
-                    <h5 class="card-title">${room.entity.propertyMap.title}</h5>
+                    <h5 class="card-title" id="room-title"></h5>
                     <a class="card-text small-text" href="#">Kyle Smith</a>
                     <div class="card-text small-text">Followers: Infinite</div>
                     <div class="card-text small-text">Tag(s): </div>
                     <div class="small-spacing-bottom"></div>
-                    <button type="button" class="btn btn-default" 
-                    onclick="window.location.href='lesson.html?room_id='+${room.entity.key.id}">View</button>
+                    <div class="btn btn-default" id="room-link"></button>
```
Comment (707090) [inline capstone/src/main/webapp/dashboard-loader.js]: \<div ...>\</button>? This doesnt freak the browser out? Most browsers are pretty forgiving, but Im surprised this works. In any case, this should be an <a> anyway (per below, changed in a separate PR)

File: capstone/src/main/webapp/lesson-loader.js [hunk 5]
```diff
@@ -6,6 +6,8 @@ const VIDEO = "video";
 const CONTENT = "content";
 const IMAGE = "image";
 
+const queryString = "/lesson?room_id=" + getRoomId() + "&action=" + getJoinStatus();
```
Comment (707090) [inline capstone/src/main/webapp/lesson-loader.js]: Terminology-wise, the query string in a url is purely the part after the '?' (and before the '#' if there is one), not the whole path. I would just call this like lessonPath or something. Good call pulling this out though

File: capstone/src/main/webapp/lesson-loader.js [hunk 6]
```diff
@@ -78,18 +80,36 @@ function getLessons() {
     });
 }
 
+function joinRoom() {
+    console.log("Room joined!");
```
Comment (707090) [inline capstone/src/main/webapp/lesson-loader.js]: Debug lines should be removed before PRs

File: capstone/src/main/webapp/lesson-loader.js [hunk 7]
```diff
@@ -78,18 +80,36 @@ function getLessons() {
...
+    return fetch("/join?room_id=" + getRoomId() + "&action=unjoin",
+    {method: "POST", headers: new Headers({ID_TOKEN})});
+}
+
 function getRoomId() {
     return new URL(window.location.href).searchParams.get("room_id");
 }
 
-// Will update in UI PR
+function getJoinStatus() {
```
Comment (707090) [inline capstone/src/main/webapp/lesson-loader.js]: This could be called getJoinAction, since "unjoining" is not a status

File: capstone/src/main/webapp/lesson-loader.js (hunk 5 above)
Comment (707090) [reply capstone/src/main/webapp/lesson-loader.js]: To make this more robust, I would use the JS classes around URLs (this includes for all requests in the app). You can write this as:

```
lessonEndpoint = new URL("/lesson", window.location.href);
lessonEndpoint.searchParams.set(ROOM_ID, getRoomId());
lessonEndpoint.searchParams.set(ACTION, getJoinAction());
```
then later when you want to use it you can fetch this instead of a string `fetch(locationEndpoint, {method...})`

**Review by 707090 — CHANGES_REQUESTED** [review]
Unjoin can also probably be called leave?

File: capstone/src/main/java/com/google/sps/servlets/JoinServlet.java (hunk 1 above)
Comment (kyleasmith1) [reply capstone/src/main/java/com/google/sps/servlets/JoinServlet.java]: gotcha

File: capstone/src/main/webapp/dashboard-loader.js (hunk 3 above)
Comment (kyleasmith1) [reply capstone/src/main/webapp/dashboard-loader.js]: yea sorry I should've made a separate PR

File: capstone/src/main/webapp/dashboard-loader.js (hunk 4 above)
Comment (kyleasmith1) [reply capstone/src/main/webapp/dashboard-loader.js]: What should this have been? Like why did you expect it to freak the browser out?

File: capstone/src/main/webapp/lesson-loader.js (hunk 5 above)
Comment (kyleasmith1) [reply capstone/src/main/webapp/lesson-loader.js]: gotcha that works perfectly

File: capstone/src/main/webapp/dashboard-loader.js (hunk 4 above)
Comment (707090) [reply capstone/src/main/webapp/dashboard-loader.js]: \<div>\</button>

Comment (kyleasmith1) [reply capstone/src/main/webapp/dashboard-loader.js]: I didn't notice and the browser didn't catch it I guess. I'll change it now so I can merge

**Review by 707090 — APPROVED** [review]
